docs/keyed_services.md

```diff
@@ -0,0 +1,153 @@
+...
+}
+}  // namespace fake
+```
+
+brave/browser/fake/fake_service_factory.cc
+```cpp
+mojo::PendingRemote<mojom::FakeService> FakeServiceFactory::GetRemoteForProfile(
+    content::BrowserContext* context) {
+  auto* instance = GetInstance()->GetServiceForBrowserContext(context, true);
+  if (!instance) {
+    return mojo::PendingRemote<mojom::FakeService>();
+  }
+  return static_cast<fake::FakeServiceImpl*>(instance)->MakeRemote();
+}
+
+void FakeServiceFactory::BindRemoteForProfile(
+    content::BrowserContext* context,
+    mojo::PendingReceiver<fake::mojom::FakeService> receiver) {
+  auto* service = static_cast<fake::FakeServiceImpl*>(
+      GetInstance()->GetServiceForBrowserContext(context, true));
+  if (service) {
+    service->Bind(std::move(receiver));
+  }
+}
+
+static jlong JNI_FakeService_GetForProfile(
+    JNIEnv* env,
+    const base::android::JavaParamRef<jobject>& profile_android) {
+  auto* profile = Profile::FromJavaObject(profile_android);
+  auto pending =
+      FakeServiceFactory::GetInstance()->GetRemoteService(profile);
+  return pending.PassPipe().release().value()
+}
+```
+
+Java
+
+brave/browser/fake/android/java/src/org/chromium/brave/browser/fake/FakeServiceFactory.java
+```java
+public @Nullable FakeService getForProfile(Profile profile,
+        @Nullable ConnectionErrorHandler connectionErrorHandler) {
+    long nativeHandle = FakeServiceFactoryJni.get().getForProfile(profile);
+    MessagePipeHandle handle =
+            CoreImpl.getInstance().acquireNativeHandle(nativeHandle).
+            toMessagePipeHandle();
+    if (!handle.isValid()) {
+      return null;
+    }
+    FakeService fakeService =
+        FakeService.MANAGER.attachProxy(handle, 0);
+    if (connectionErrorHandler != null) {
+        Handler handler = ((Interface.Proxy) fakeService).getProxyHandler();
+        handler.setErrorHandler(connectionErrorHandler);
+    }
+    return fakeService;
+}
+```
+
+#### IOS
+
+IOS has separate factories because Profile subclasses `web::BrowserState` instead of `content::BrowserContext`, but the same principles apply to `ProfileKeyedServiceFactoryIOS`.
+
+#### IOS and Desktop profile checks
+
+TODO - standardize a way to maintain consistency between `GetBrowserContextToUse` methods on desktop and android. Possibly using some constants like `kRegular`, `kIncognito`, `kTor` so the same code can be used to check both `Profile` and `ProfileIOS`
+
+#### IOS keyed services and mojo
+
+obj-c
+
+brave/ios/browser/fake/fake_service_factory.mm
+```cpp
+mojo::PendingRemote<fake::mojom::FakeService> FakeServiceFactory::GetForProfile(
+    ProfileIOS* profile) {
+  auto* service = GetInstance()->GetServiceForProfileAs<fake::FakeServiceImpl>(
+      profile, true);
+  if (!service) {
+    return mojo::PendingRemote<fake::mojom::FakeService>();
+  }
+  return service->MakeRemote();
+}
+
+@implementation FakeFakeServiceFactory
++ (nullable id)getForProfile:(ProfileIOS*)profile {
```

Comment:
`serviceForProfile`

Reply:
actually I think this should match what we have everywhere else moving forward

Reply:
I think ideally we actually want to wrap & export `ProfileIOS` and just expose `getForProfile` directly instead of having `getForPrivateBrowsing` and `serviceForProfile`, but we can rename `serviceForProfile` in the interim if we want

Reply:
but this method does take ProfileIOS and I copied it from somewhere... I guess I should create a little PR and test that all this code is correct

Reply:
Yeah because `getForPrivateBrowsing` is just a helper method that calls `serviceForProfile` with the correct profile